egs/wsj/s5/steps/nnet3/lstm/make_configs.py

```diff
@@ -170,17 +188,42 @@ def ParseSpliceString(splice_indexes, label_delay=None):
-        prev_layer_output = nodes.AddLstmLayer(config_lines, "Lstm{0}".format(i+1), prev_layer_output, args.cell_dim,
-                                         args.recurrent_projection_dim, args.non_recurrent_projection_dim,
-                                         args.clipping_threshold, args.norm_based_clipping,
-                                         args.ng_per_element_scale_options, args.ng_affine_options,
-                                         lstm_delay = lstm_delay[i])
+	if len(lstm_delay[i]) == 2: # BLSTM layer case, add both forward and backward
+	    try:
+	        if not (lstm_delay[i][0] < 0 and lstm_delay[i][1] > 0):
+	            raise ValueError('Warning: ' + str(lstm_delay[i]) + ' is not a standard BLSTM mode. There should be a negative delay for the forward, and a postive delay for the backward.')
+	    except ValueError as e:
+		print(str(e))
+            prev_layer_output1 = nodes.AddLstmLayer(config_lines, "Lstm{0}f".format(i+1), prev_layer_output, args.cell_dim,
+                                             args.recurrent_projection_dim, args.non_recurrent_projection_dim,
+                                             args.clipping_threshold, args.norm_based_clipping,
+                                             args.ng_per_element_scale_options, args.ng_affine_options,
+                                             lstm_delay = lstm_delay[i][0])
+            prev_layer_output2 = nodes.AddLstmLayer(config_lines, "Lstm{0}b".format(i+1), prev_layer_output, args.cell_dim,
+                                             args.recurrent_projection_dim, args.non_recurrent_projection_dim,
+                                             args.clipping_threshold, args.norm_based_clipping,
+                                             args.ng_per_element_scale_options, args.ng_affine_options,
+                                             lstm_delay = lstm_delay[i][1])
+            prev_layer_output['descriptor'] = 'Append({0}, {1})'.format(prev_layer_output1['descriptor'], prev_layer_output2['descriptor'])
+	    prev_layer_output['dimension'] = prev_layer_output1['dimension'] + prev_layer_output2['dimension']
+	else: # LSTM layer case
+	    prev_layer_output = nodes.AddLstmLayer(config_lines, "Lstm{0}".format(i+1), prev_layer_output, args.cell_dim,
+			                    args.recurrent_projection_dim, args.non_recurrent_projection_dim,
+					    args.clipping_threshold, args.norm_based_clipping,
+					    args.ng_per_element_scale_options, args.ng_affine_options,
+					    lstm_delay = lstm_delay[i][0])
         # make the intermediate config file for layerwise discriminative
         # training
         nodes.AddFinalLayer(config_lines, prev_layer_output, args.num_targets, args.ng_affine_options, args.label_delay)
         config_files['{0}/layer{1}.config'.format(args.config_dir, i+1)] = config_lines
         config_lines = {'components':[], 'component-nodes':[]}
+	if len(lstm_delay[i]) == 2:
```

Comment:
Clearly describe why you are doing this descriptor modification. I remember that you needed it as you could cascade Append descriptors. It is better that you mention it here for the sake of clarity.